File: day14/disk_defrag.pl
#!/usr/bin/env perl
use strict;
use warnings;
use feature qw(:5.24);
use experimental qw(signatures);
use List::Util qw(sum);
use Graph::Undirected;

my $key = $ARGV[0];
my $total = 0;
my @disk;
for my $i (0..127) {
    my $hash = knot_hash("$key-$i");
    $total += sum @$hash;
    push @disk, $hash;
}

say "result1: $total";

my $g = Graph::Undirected->new;
# add vertex for each used square
for my $r (0..127) {
    for my $c (0..127) {
        $g->add_vertex("$r,$c") if $disk[$r][$c];
    }
}

# rows
for my $r (0..127) {
    for my $c (0..126) {
        if ($disk[$r][$c] && $disk[$r][$c+1]) {
            my $from = "$r,$c";
            my $to = sprintf "%d,%d", $r, $c+1;
            $g->add_edge($from, $to);
        }
    }
}
# cols
for my $r (0..126) {
    for my $c (0..127) {
        if ($disk[$r][$c] && $disk[$r+1][$c]) {
            my $from = "$r,$c";
            my $to = sprintf "%d,%d", $r+1, $c;
            $g->add_edge($from, $to);
        }
    }
}
my @cc = $g->connected_components();
say "result2: ", scalar @cc;

sub knot_hash($lens) {

    my @list = 0..255;
    my @STD_SUFFIX = (17, 31, 73, 47, 23);
    my @lens = map {ord} split //, $lens;
    push @lens, @STD_SUFFIX;

    my $p = 0;
    my $skip = 0;

    for my $round (1..64) {
        for my $len (@lens) {
            my $end = $p + $len - 1;
            if ($end <= $#list) {
                @list[$p..$end] = reverse @list[$p..$end];
            } else {
                my @tmp = (@list, @list);
                @tmp[$p..$end] = reverse @tmp[$p..$end];
                @tmp[0..($end-$#list-1)] = @tmp[$#list+1..$end];
                @list = @tmp[0..$#list];
            }
            $p = ($p + $len + $skip) % @list;
            $skip++;
        }
    }

    # compute dense hash from @list
    my @dense;
    for (my $i = 0; $i < 256; $i += 16) {
        my $out = $list[$i];
        for my $j ($i+1..$i+15) {
            $out ^= $list[$j];
        }
        push @dense, $out;
    }
    my $hash;
    $hash .= sprintf "%08b", $_ for @dense;
    return [split //, $hash];
}

File: day14/disk_defrag.py
#/usr/bin/env python3
from sys import argv
import networkx as nx

def knot_hash(lens_str):
    myList = [x for x in range(0,256)]
    STD_SUFFIX = [17, 31, 73, 47, 23]
    lens = [ord(c) for c in lens_str] + STD_SUFFIX

    p = 0
    skip = 0

    for round in range(1,65):
        for myLen in lens:
            end = p + myLen - 1
            if end < len(myList):
                slice = myList[p:end+1]
                myList[p:end+1] = slice[::-1]
            else:
                tmp = myList + myList
                slice = tmp[p:end+1]
                tmp[p:end+1] = slice[::-1]
                tmp[0:(end-len(myList)+1)] = tmp[len(myList):end+1]
                myList = tmp[0:len(myList)]

            p = (p + myLen + skip) % len(myList)
            skip += 1

    # compute dense hash from myList
    dense = []
    for i in range(0,256,16):
        out = myList[i]
        for j in range(i+1,i+16):
            out ^= myList[j]
        dense.append(out)

    hash = []
    for val in dense:
        bits = "{:08b}".format(val)
        hash += [int(x) for x in bits]
    return hash

key = argv[1]
total = 0
disk = []
for i in range(128):
    keyi = "{}-{:d}".format(key, i)
    hash = knot_hash(keyi)
    total += sum(hash)
    disk.append(hash)

print("result1:", total)

g = nx.Graph()

# add vertex for each used square
for r in range(128):
    for c in range(128):
        if disk[r][c]:
            key = "{},{}".format(r,c)
            g.add_node(key)

# rows
for r in range(128):
    for c in range(128-1):
        if disk[r][c] and disk[r][c+1]:
            fromKey = "{},{}".format(r,c)
            toKey = "{},{}".format(r,c+1)
            g.add_edge(fromKey, toKey)
            
# cols
for r in range(128-1):
    for c in range(128):
        if disk[r][c] and disk[r+1][c]:
            fromKey = "{},{}".format(r,c)
            toKey = "{},{}".format(r+1,c)
            g.add_edge(fromKey, toKey)
            
cc = list(nx.connected_components(g))
print("result2:", len(cc))
